std/process.d

```diff
@@ -1042,56 +1043,102 @@ private Pid spawnProcessPosix(scope const(char[])[] args,
+                        import core.stdc.stdlib : malloc;
+                        import core.sys.posix.poll : pollfd, poll, POLLNVAL;
+                        import core.sys.posix.sys.resource : rlimit, getrlimit, RLIMIT_NOFILE;
+
+                        // Get the maximum number of file descriptors that could be open.
+                        rlimit r;
+                        if (getrlimit(RLIMIT_NOFILE, &r) != 0)
+                        {
+                            abortOnError(forkPipeOut, InternalError.getrlimit, .errno);
+                        }
+                        immutable maxDescriptors = cast(int) r.rlim_cur;
+
+                        immutable maxToClose = maxDescriptors - lowfd;
+
+                        // Call poll() to see which ones are actually open:
+                        auto pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
+                        if (pfds is null)
+                        {
+                            abortOnError(forkPipeOut, InternalError.malloc, .errno);
+                        }
+                        foreach (i; 0 .. maxToClose)
+                        {
+                            pfds[i].fd = i + lowfd;
+                            pfds[i].events = 0;
+                            pfds[i].revents = 0;
+                        }
+                        if (poll(pfds, maxToClose, 0) >= 0)
+                        {
+                            foreach (i; 0 .. maxToClose)
+                            {
+                                // POLLNVAL will be set if the file descriptor is invalid.
+                                if (!(pfds[i].revents & POLLNVAL)) close(pfds[i].fd);
+                            }
+                        }
+                        else
+                        {
+                            // Fall back to closing everything.
+                            foreach (i; lowfd .. maxDescriptors)
+                            {
+                                close(i);
+                            }
+                        }
                     }
-                }
-                else
-                {
-                    // Fall back to closing everything.
-                    foreach (i; 3 .. maxDescriptors)
-                    {
-                        if (i == forkPipeOut) continue;
-                        close(i);
+
+                    // closefrom may not be available on the version of glibc we build against.
+                    // Until we find a way to perform this check we will try to use dlsym to
+                    // check for the function. See: https://github.com/dlang/phobos/pull/9048
+                    version (CRuntime_Glibc) {
+                        void closefrom (int lowfd) {
+                            static bool tryGlibcClosefrom (int lowfd) {
+                                import core.sys.posix.dlfcn;
+
+                                void *handle = dlopen("libc.so.6", RTLD_LAZY);
```

Comment:
1. Why `libc.so.6` and not `libc.so`?
2. What do you think about using `RTLD_NOLOAD`?

Reply:
> 1. Why `libc.so.6` and not `libc.so`?

If `dlopen("libc.so")` ever finds something different than `dlopen("libc.so.6")` then we definitely don't want to continue trying to load symbols. Very unlikely that this will ever happen but better safe then sorry.

> 2. What do you think about using `RTLD_NOLOAD`?

Looking only at the man page of `dlopen` I can't tell what those values actually do so I went with the value they showed in the example.

Reply:
My thinking is that in the unlikely case that the current binary is actually using a libc that's not `libc.so.6` then we probably don't want to actually load a second `libc`. I understand that `RTLD_NOLOAD` should achieve that.

Reply:
`RTLD_NOLOAD` makes it so that `closefrom` is no longer found so I'll keep `RTLD_LAZY`.

Reply:
1. This debate could be solved by avoiding dlopen/dlsym
2. I find it vanishingly unlikely in the event Cruntime_Glibc is true but the soversion of glibc has advanced past "6", that you'll *also* have libc.so.6 available